[feature] 스터디 공고 지원 - 지원 내역 목록 조회

Jiwoong0322 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #134 
- 작업 요약: 스터디 공고 지원 - 지원 내역 목록 조회 기능 개발

## 📄 상세 내용
- [ ] 공고 지원자 목록 조회 API 구현 (GET /api/v1/recruitments/{recruitment_uuid}/applications/all)
- [ ] 커서 기반 페이지네이션(DefaultCursorPagination) 기능 적용
- [ ]  Application 모델의 str 메소드 동작을 검증하는 단위 테스트 추가
- [ ]  공고 작성자만 목록을 조회할 수 있도록 IsRecruitmentAuthor 권한 추가

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Jiwoong0322 committed:
✨ feat: 스터디공고 지원 - 지원 내역 목록 조회

  1.  공고 지원자 목록 조회 API 구현 (GET /api/v1/recruitments/{recruitment_uuid}/applications/all)
  2.  공고 작성자만 목록을 조회할 수 있도록 IsRecruitmentAuthor 권한 추가
  3.  커서 기반 페이지네이션(DefaultCursorPagination) 기능 적용
  4.  Application 모델의 __str__ 메소드 동작을 검증하는 단위 테스트 추가

관련 이슈: #134

Jiwoong0322 committed:
♻️ refactor: applications 파일 통합

1. API 엔드포인트 및 View 구조 개선
 - 지원서 생성와 지원자 목록 조회 API의 엔드포인트 통합
 - 분리되어 있던 View들을 ApplicationAPIView 하나로 통합.
2. 권한 로직 최적화
 - IsRecruitmentAuthor 권한 클래스에서 공고 객체 조회 시, 해당 객체를 request에 저장하도록 하여 View에서 발생하는 중복 DB 조회를 방지하고 성능을 개선
3. 코드 구조 리팩토링 및 정리
 - 여러 개로 분리되어 있던 managers를 하나로 통합, 리팩토링 과정에서 불필요해진 view, manager, service 파일을 삭제
4. 테스트 및 타입 에러 수정
 - 리팩토링으로 인해 변경된 URL 이름, 권한 로직, 누락됨 import등 테스트 코드와 소스 코드에 반영하여 mypy 타입 에러와 테스트 실패 해결.

관련 이슈: #134

# Conflicts:
#	apps/recruitments/urls/recruitments_urls.py

# Conflicts:
#	apps/recruitments/urls/recruitments_urls.py

# Conflicts:
#	apps/applications/views/applications_views.py

Jiwoong0322 committed:
♻️ refactor: mypy 에러 해결

1. applications_views.pydml get_permissions 함수에 반환 타입 어노테이션이 누락되있던 문제 해결
2. 들여쓰기 오류 해결

관련 이슈: #134

Umdoong commented on apps/applications/tests/test_applications_list.py:
```diff
@@ -0,0 +1,164 @@
+    REQ-APLY-002: 본인이 작성한 공고에 대한 지원 목록 조회 API 테스트
+    """
+
+    def setUp(self) -> None:
```
우리 지난주 엔지니어링 데이에서 setUpTestData 배웠잖아요...!ㅠ

Umdoong commented on apps/recruitments/recruitments_permissions.py:
```diff
@@ -0,0 +1,23 @@
+    - DB 중복 조회를 막기 위해 조회된 recruitment 객체를 request에 저장
+    """
+
+    def has_permission(self, request: Request, view: APIView) -> bool:
```
객체에 대한 권한을 검사하는 거라서 has_permission이 아니라 has_object_permission을 오버라이드 하셔야 할 것 같습니다.

Jiwoong0322 committed:
♻️ refactor: API 응답 필드 추가

API 응답 시 참여 신청 id(application_id)를 포함하여 응답하도록 함.

관련 이슈: #134

Jiwoong0322 committed:
♻️ refactor: API 응답 변경 테스트 적용

API 응답 시 application_id를 추가해서 응답하도록 변경한 내용을 테스트코드에 추가

관련 이슈: #134

Umdoong commented on apps/applications/views/applications_views.py:
```diff
@@ -0,0 +1,75 @@
+            user = cast(User, request.user)
+            service = ApplicationService()
+            application = service.create_application(user=user, recruitment=recruitment, **serializer.validated_data)
+        except IntegrityError as e:
```
integrity error는 어플리케이션의 중복 때만 발생하는 게 아니기 때문에 중복의 경우 if문을 통해서 직접 중복 여부를 확인하는 게 더 좋아보입니다.
그리고 에러메세지는 {"error": "에러메세지"}로 통일하고 있으니 맞춰주세요!

Jiwoong0322 committed:
♻️ refactor: 지원 API 권한 및 테스트 로직 수정

  1. IsRecruitmentAuthor 권한 클래스 로직 수정 (has_object_permission 사용)
  2. 중복 지원 검증 로직 및 에러 메시지 형식 통일
  3. 테스트 효율성을 위한 setUpTestData 적용 및 mypy 타입 에러 해결

관련 이슈: #134

Meoyoug reviewed (approved):
고생하셨습니다 머지할게용
